src/store/data-store-level.ts

```diff
@@ -50,27 +50,25 @@ export class DataStoreLevel implements DataStore {
     await this.blockstore.close();
   }
 
-  async put(tenant: string, messageCid: string, dataStream: Readable): Promise<PutResult> {
-    const dataPartition = await this.blockstore.partition(DATA_PARTITION);
+  async put(tenant: string, messageCid: string, dataCid: string, dataStream: Readable): Promise<PutResult> {
+    const blocksForData = await this.blockstore.partition(DATA_PARTITION);
+    const blocks = await blocksForData.partition(dataCid);
 
-    const asyncDataBlocks = importer([{ content: dataStream }], dataPartition, { cidVersion: 1 });
+    const asyncDataBlocks = importer([{ content: dataStream }], blocks, { cidVersion: 1 });
 
     // NOTE: the last block contains the root CID as well as info to derive the data size
     let block: ImportResult;
     for await (block of asyncDataBlocks) { ; }
 
-    const dataCid = block.cid.toString();
-    const dataSize = block.unixfs ? Number(block.unixfs!.fileSize()) : Number(block.size);
-
     const tenantsForData = await this.blockstore.partition(HOST_PARTITION);
     const messagesForTenant = await tenantsForData.partition(dataCid);
     const messages = await messagesForTenant.partition(tenant);
 
     await messages.put(messageCid, PLACEHOLDER_VALUE);
 
     return {
-      dataCid,
-      dataSize
+      dataCid  : block.cid.toString(),
```

Comment:
perhaps it was a mistake on my part also calling this `dataCid`.  i almost see it more as a "give this big blob of data a name" rather than "what is the actual `CID` of this data"

im not necessarily against moving the exceptions into this, but i think the problem there is that if the developer provides their own `DataStore` then suddenly we'd lose these safety checks (and along those lines id kinda rather `DataStoreLevel` assume everyone is a good actor so that it's logic can stay clean)

Reply:
Agree with you that it'd be ideal that we can lift any burden away from the data store implementation AMAP, but here we are, requiring `dataCid` as part of the interface mainly because level impl needs it, this kinda is going against that spirit, so it was sorta of a tradeoff situation. In the same spirit, the "bypassing data stream if data of CID exists check" logic would also be better to be outside the data store implementation?

I want to be agreeable to the PR and there are better hills to die on. If you can add the documentation of expectation of `dataCid` param in the `put` method signature of the `Datastore` interface (along with the condition for ignoring data stream if required), then we can call this resolved.

The original thinking was that a "thorough/defensive" programmer would end up doing the CID check as it is odd to not check it given it is the expected `dataCid` after all. Which then begs the question why have it in two places if they were gonna check it.

Reply:
yeah i suppose it's somewhat expected that a developer building their own `DataStore` would do the right thing if given both a `dataCid` and `dataSize` (in addition to the `dataStream`) so i suppose we could move the checks inside `DataStoreLevel`.  put another way, there's no guarantee the developer wouldnt lie when returning the `dataCid` or `dataSize` (though if we dont provide the `dataSize` as an argument then they'd at least have to get that part right, which i suppose was part of the idea when neither `dataCid` or `dataSize` were provided)

lemme try moving the exceptions into `DataStoreLevel` and see how it looks/feels 🤔

Reply:
yeah IMO it felt kinda funky to put it in the `DataStoreLevel`, especially since the check for `associate` was not, which made me want to put that check inside too

instead i felt it was better to unify everything and have all the methods return both a `dataCid` and `dataSize` for verification at every step